pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,426 @@
+        FailedToPayProtocolFee,
+        /// Failed To charge the instantiation fee for the smart extension.
+        FailedToPayInstantiationFee,
+        /// Given identityId is not CDD.
+        NewOwnerIsNotCDD,
+        /// Insufficient max_fee provided by the user to instantiate the SE.
+        InsufficientMaxFee,
+    }
+}
+
+decl_event! {
+    pub enum Event<T>
+        where
+        Balance = BalanceOf<T>,
+        CodeHash = <T as frame_system::Trait>::Hash,
+    {
+        /// Emitted when instantiation fee of a template get changed.
+        /// IdentityId of the owner, Code hash of the template, old instantiation fee, new instantiation fee.
+        InstantiationFeeChanged(IdentityId, CodeHash, Balance, Balance),
+        /// Emitted when the instantiation of the template get frozen.
+        /// IdentityId of the owner, Code hash of the template.
+        InstantiationFreezed(IdentityId, CodeHash),
+        /// Emitted when the instantiation of the template gets un-frozen.
+        /// IdentityId of the owner, Code hash of the template.
+        InstantiationUnFreezed(IdentityId, CodeHash),
+        /// Emitted when the template ownership get transferred.
+        /// IdentityId of the owner, Code hash of the template, IdentityId of the new owner of the template.
+        TemplateOwnershipTransferred(IdentityId, CodeHash, IdentityId),
+        /// Emitted when the template usage fees gets changed.
+        /// IdentityId of the owner, Code hash of the template,Old usage fee, New usage fee.
+        TemplateUsageFeeChanged(IdentityId, CodeHash, Balance, Balance),
+        /// Emitted when the template instantiation fees gets changed.
+        /// IdentityId of the owner, Code hash of the template, Old instantiation fee, New instantiation fee.
+        TemplateInstantiationFeeChanged(IdentityId, CodeHash, Balance, Balance),
+        /// Emitted when the template meta url get changed.
+        /// IdentityId of the owner, Code hash of the template, old meta url, new meta url.
+        TemplateMetaUrlChanged(IdentityId, CodeHash, Option<MetaUrl>, Option<MetaUrl>),
+    }
+}
+
+decl_module! {
+    // Wrap dispatchable functions for contracts so that we can add additional gating logic.
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        /// Initialize the default event for this module.
+        fn deposit_event() = default;
+
+        /// Error type.
+        type Error = Error<T>;
+
+        /// The minimum amount required to generate a tombstone.
+        const NetworkShareInInstantiationFee: Perbill = T::NetworkShareInFee::get();
+
+        // Simply forwards to the `update_schedule` function in the Contract module.
+        #[weight = pallet_contracts::Call::<T>::update_schedule(schedule.clone()).get_dispatch_info().weight]
+        pub fn update_schedule(origin, schedule: Schedule) -> DispatchResult {
+            <pallet_contracts::Module<T>>::update_schedule(origin, schedule)
```

Comment:
We can factor in
```
pallet_permissions::Module<T>::ensure_call_permissions(&sender)?;
```
if that makes sense. Not sure if you wanted a simple redirection or maybe we could gate it.